### ISSUE TITLE
store metadata as their original type instead of casting it all to string, and guess-cast it back

### ISSUE THREAD

**Issue by carmenfan** [3drepo/3drepobouncer#462]
### Description
At the moment the metadata constructor takes in strings for the values. Then use boost to guess their original types and cast them back.

This is a bit backwards as very often we're [casting the original type to string](https://github.com/3drepo/3drepobouncer/blob/master/bouncer/src/repo/manipulator/modelconvertor/import/repo_model_import_3drepo.cpp#L63) in order to fit the function definition here, only to cast it back to the [original type](https://github.com/3drepo/3drepobouncer/blob/master/bouncer/src/repo/core/model/bson/repo_bson_factory.cpp#L223):

It's probably better if we use a variant type (like https://en.cppreference.com/w/cpp/utility/variant ) to keep the type until the very end.

### Goal
- Create a `RepoVariant` class, inheriting std library variant type
- Update the metadata constructor to take in `RepoVariant for value type instead of strings
  - Remove the lexical cast attempts
- Update all importers with metadata to stop the casting their data to string
  - If there are any importers that are providing the value as string already, attempt the lexical cast there

**Comment by FJThiel** [comment]
@sebjf @carmenfan 
I reviewed the different importers and suggest the following:

## The Variant Itself
The variant needs to contain at least the types long long, double, and std::string, since these are the three types that we guesstimate our strings into currently. Since the variant allows us to carry more types, we could be more granular if we want (e.g. differentiate between a char, int, or long), but the current bouncer manages with those three, so that is the minimum.

## Getting Data Out of the Variant
I think the obvious place to convert the variant back into regular data types is in the RepoBSONBuilder in the [append(...)](https://github.com/3drepo/3drepobouncer/blob/96c4317e3d43fdcaab9cf6ed100dc895e4d0544b/bouncer/src/repo/core/model/bson/repo_bson_builder.h#L77) method. This is the last location before we hand the data over to the mongo's BSON builder. Anything before that does not need to know what is actually in the variant and it comes with the additional benefit of making append more robust. Right now, it is a template method that accepts any class but mongo's append is not a templace, so calling our append with the wrong type as T will break. If our append only accepts our variant, that won't be possible anymore.

## Getting Data in the Variant
This will require individual treatment for each importer.

### 3D Repo Model Importer
The conversion of the meta data into strings happens in [RepoModelIMport::createMetadataNode](https://github.com/3drepo/3drepobouncer/blob/96c4317e3d43fdcaab9cf6ed100dc895e4d0544b/bouncer/src/repo/manipulator/modelconvertor/import/repo_model_import_3drepo.cpp#L45). They are pulled out of a boost ptree, which comes from a json structure. Changing the JSON to already carry the variant is probably going a bit far, so I suggest that the variant is created and filled with the data right where we are currently converting it into strings.

### Assimp Importer
The conversion to strings happens in [AssimpModelImport::createMetadataRepoNode.](https://github.com/3drepo/3drepobouncer/blob/96c4317e3d43fdcaab9cf6ed100dc895e4d0544b/bouncer/src/repo/manipulator/modelconvertor/import/repo_model_import_assimp.cpp#L551) There, an aiMetadataEntry object contains the meta data. This object has a void* containing the actual data and an enum to designate the type. It is practically a poor man's less typesafe variant. I suggest that we convert it right there into our variant. Ideally with a dedicated constructor for our variant.

### IFC Importer 
With IFC, the conversion happens in [TreeParser::determineActionsByElementType](https://github.com/3drepo/3drepobouncer/blob/96c4317e3d43fdcaab9cf6ed100dc895e4d0544b/ifcUtils/repo_ifc_utils_tree_parser.cpp#L241). There, the element passed as a parameter is checked by trying to dynamic cast it into different subtypes. Some of them are types that produce meta data. If one of them is found, the data is extracted in its native type and then converted to string, currently. I suggest that it instead is put into our variant.

### ODA
The ODA import is split into four different data processors depending on the format.

#### DGN Data Processor
The data is pulled out of the XML structure in [DataProcessorDgn::extractXMLLinkages](https://github.com/3drepo/3drepobouncer/blob/96c4317e3d43fdcaab9cf6ed100dc895e4d0544b/bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/data_processor_dgn.cpp#L61). Unfortunately, the class that does it does not appear to be under our control and returns only a string. I fear we can't do more here than doing a lexical cast here before putting it into our variant.

#### DWG Data Processor
This is an odd one. It does not appear to have much meta data. All that I found was in [DataProcessorDwg::doDraw](https://github.com/3drepo/3drepobouncer/blob/96c4317e3d43fdcaab9cf6ed100dc895e4d0544b/bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/data_processor_dwg.cpp#L34) and consists of strings representing handles being stored with the key "Entity Handle::Value". If that is indeed all, we can just add the variant to those lines and that is it.

#### NWD Data Processor
As if to compensate for DWG, there is a lot of meta data here, which is all processed in [DataProcessorNwd::processAttributes](https://github.com/3drepo/3drepobouncer/blob/96c4317e3d43fdcaab9cf6ed100dc895e4d0544b/bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/data_processor_nwd.cpp#L220). The data comes in via an OdNwPropertyPtr which can be queried about its type before retrieving the data from it. I suggest to convert directly from that type to our variant similarly to the way I suggest for Assimp.

#### RVT Data Processor
Similar to the NWD Data Processor, this too comes with its own type: OdTfVariant. This is no std::variant though. Instead, it is a class encapsulating a union to make it typesafe. As with Assimp and NWD, I suggest to directly convert the type into our variant. This would be done in [DataProcessorRvt::translateMetadataValue.](https://github.com/3drepo/3drepobouncer/blob/96c4317e3d43fdcaab9cf6ed100dc895e4d0544b/bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/data_processor_rvt.cpp#L91)

### Synchro Importer
The data from the synchro importer already enters our scope as strings, so doing things on the bouncer side would be working with lexical casts again. However, we do have the possibility to change things in our spm reader library. If [SPMReader::getAllMetadata](https://github.com/3drepo/SynchroReader/blob/9a9ef980899b88651a2c4039d0ee58aa38971a92/plugins/common/spm_reader.cpp#L233) can get changed so that it returns our variant (or some intermediary format), this can be directly be used on the side of the bouncer without converting to strings and back.

**Comment by carmenfan** [comment]
@FJThiel 

> The variant needs to contain at least the types long long, double, and std::string, since these are the three types that we guesstimate our strings into currently. Since the variant allows us to carry more types, we could be more granular if we want (e.g. differentiate between a char, int, or long), but the current bouncer manages with those three, so that is the minimum.

Can we add boolean to the list? the only reason we don't do it right now is because boost can't tell the difference 🙈 

I don't see we'd ever want char, or long. but if possible it is possibly good to distinguish int from double so we can avoid underflows (you never know!)

**Comment by FJThiel** [comment]
@carmenfan 

> Can we add boolean to the list? the only reason we don't do it right now is because boost can't tell the difference 🙈
> 
> I don't see we'd ever want char, or long. but if possible it would be good to distinguish int from double so we can avoid underflows (you never know!)

Okay, bools are added to the list!
So we carry bool, int, double, and string?

> I would just keep the lexical cast for now. Considering we want to do this as a stepping stone to remove mongo dependency and upgrade the mongodb driver, it's not a major concern if synchro importer is still doing the old ways. (I want to spare you from looking at another terrible library... 😆 )

Okay, I will keep it as is and do the lexical cast for synchro.

**Comment by carmenfan** [comment]
> So we carry bool, int, double, and string?

Might need to see what you get from the importer, but it might be useful to be able to store date as mongo date types too. 

but essentially yes I think so!

**Comment by FJThiel** [comment]
We might have a problem with our toolchain for bouncer. v140 has no full C++17 support. According to MS, variants are only VS 2017 and up
https://devblogs.microsoft.com/cppblog/announcing-msvc-conforms-to-the-c-standard/

**Comment by FJThiel** [comment]
However, if I understand [this](https://learn.microsoft.com/en-us/cpp/porting/binary-compat-2015-2017?view=msvc-170) correctly, we might get around re-compiling everything. There appears to be some compatibility between the different version of the toolsets. I will try this.

**Comment by carmenfan** [comment]
The other option is to just use boost version of the variant. Sebastian tried to update to c++17 and encountered some pain iirc, so you might need to check with him whether you'll run into more issues than just recompiling 😆 

We use ubuntu with gcc v9.x in the container so I think this is only an issue amongst our machines. If we do end up upgrading the toolchain, drop the dlls onto sharepoint please 😄

**Comment by sebjf** [comment]
@carmenfan, @FJThiel 

> Sebastian tried to update to c++17 and encountered some pain iirc

I was just writing this up!

IIRC, bouncer will compile with C++17 - basically when we did the decimation stuff, we needed c++17 libraries, so I both fixed up bouncer for c++17, and made the libraries work with polyfills so we could stay on 14. I can check the issue again but my instincts are that some of the difficulties were with ODA, but we've upgraded that since.

On a hunch I checked out Mongo's documentation, and sure enough the new driver prefers c++17 too, and requests polyfills for previous toolchains: 

https://www.mongodb.com/docs/languages/cpp/cpp-driver/current/polyfill-selection/

The reason we didn't upgrade the toolchain back then, was because with the polyfills for the decimation libraries, there was just no need, and we didn't want to risk breaking anything.

However, with both this and Mongo, it may now be worth it.

As @FJThiel notes above, on the msvc toolchain at least the old binaries should continue to be linked successfully. For Linux, I believe we use the latest gcc on Ubuntu 18, and just set the feature set via the std flag so there should be no binary incompatibilities.

My personal feelings are that we are going to encounter this more and more going forward, so as long as the existing dependencies remain ABI compatible, we should upgrade the c++ version of bouncer itself.

I will look at the issue for the decimation stuff and recall what exactly were the issues with 17 back then in either case for completeness.

**Comment by FJThiel** [comment]
@sebjf @carmenfan 

I can confirm that Boost is indeed very unhappy with the switch. The binaries might be fine (at least I am not getting a compiler error for them), but the boost headers that are included are very much a problem. At least one is using features that have been deprecated in C++ 17, so I can't get build to complete.

Looks like if we would want to upgrade to C++ 17 we would also have to upgrade boost or get rid of it. Whichever is more feasible.

**Comment by sebjf** [comment]
Which boost headers are causing an issue? Is it to do with the filesystem?

**Comment by FJThiel** [comment]
Looks like everything that uses std::auto_ptr, which is apparently a lot. I don't see anything related to the file system though, but it is around 4k errors, so it might have slipped through.

It is not only boost as well. There is some mongo stuff here too (bsonmisc.h, dbclientinferface.h, dbclientcursor.h, etc.).

**Comment by sebjf** [comment]
That sounds familiar, infuriatingly I can't remember what the exact problem was though! I recall having to add a shim somewhere. Can you post some more details about the problem specifically with regards to filesystem. I can also check out #599 and look for that header.

**Comment by FJThiel** [comment]
I was just assembling a list of boost headers that have issues. Attached below. From what I can tell, it boils down to three things they use that are deprecated in C++ 17:
[std::auto_ptr](https://en.cppreference.com/w/cpp/memory/auto_ptr)
[std::unary_function](https://en.cppreference.com/w/cpp/utility/functional/unary_function)
[std::binary_function](https://en.cppreference.com/w/cpp/utility/functional/binary_function)

Headers with issues:
shared_count.hpp
shared_ptr.hpp
get_pointer.hpp
hash.hpp
case_conv.hpp
chain.hpp
compare_pointees.hpp
converter.hpp
data_extractor.hpp
extensions.hpp
function_template.hpp
get_pointer.hpp
scoped_ptr.hpp
util.hpp

**Comment by FJThiel** [comment]
Filesystem appears to be only partially involved since it uses the boost::shared_ptr.

**Comment by carmenfan** [comment]
We don't use auto_ptr anywhere, but mongodb library does. So we'd have to do #83 before we can get rid of all that 🙈 

If the only need for C++17 right now is the std variant class, then I'd suggest we use the boost variant instead for this issue, and have a separate maintenance ticket to upgrade to C++17 and we can tackle that after #83

**Comment by carmenfan** [comment]
and there's probably a separate ticket we should consider adding, where we remove any boost dependencies that are now part of the std 🙈

**Comment by FJThiel** [comment]
Sounds like a sensible plan to me. @sebjf ?

**Comment by sebjf** [comment]
> We don't use auto_ptr anywhere, but mongodb library does. 

😭 

Yes this seems most straighforward; @FJThiel , how close is the boost variant to the std version?

The only tweak I'd make to the plan is for the C++ upgrade to occur with #83, so we don't have to add a polyfill in the meantime (since its probably going to be a pain, and OS specific too...)

Boost seem to have a flag for the unary functions (BOOST_NO_CXX98_FUNCTION_BASE)

**Comment by FJThiel** [comment]
@sebjf 
There are differences in implementation but, given our relatively simple use case, I don't think they will affect us.

https://stackoverflow.com/questions/40201371/what-are-the-differences-between-stdvariant-and-boostvariant

**Comment by FJThiel** [comment]
@sebjf @carmenfan 
I noticed that there are also vectors in the meta data coming out of the assimp converter. With the current behaviour they are just stuffed into a string and never converted back i.e. handed off to mongo still as a string.

I have now the chance to have that handled properly, but I think that should also mean altering whatever reads that meta data since it will expect the vector as a string, right?

**Comment by carmenfan** [comment]
Not really worth the effort, if we consider the context here these are metadata that is being grabbed here is just going to be displayed in the attribute data card, so the only usage around the data is smart group query and... customers eye balling it

since smart group query does not support vector types it's not really going to do us much good other than having a harder time for the front end guys to represent it in the UI.

Also I don't think we get vector data on any other format, and Assimp isn't used for any of the BIM data types so we actually don't care that much about Assimp's metadata 🙈

**Comment by FJThiel** [comment]
Just putting this here in case the problem arises when the bouncer is compiled with GCC:

The authors of boost::variant warn that due to the way it is implemented, the compiler may run out of memory or its template instantiation depth may be exceeded. It works just fine for me without any additional tinkering, but if it causes issues, this documentation page mentions some flags to get around the issue:
https://www.boost.org/doc/libs/1_85_0/doc/html/variant/misc.html#variant.troubleshooting.compiler-memory

**Comment by FJThiel** [comment]
@carmenfan 

With the floating precision values not being converted to string and back, we get more decimal places in the frontend (see pictures below). Sebastian thinks the user might not expect that many decimals and that maybe adding a teamspace setting would be justified to cut them off at a user-defined length. What do you think?

After the changes:
![image](https://github.com/user-attachments/assets/e6b7e426-b7bb-4eba-9191-75bb50bd1bfa)

Before the changes:
![image](https://github.com/user-attachments/assets/a82e504e-aa7e-4566-8187-e30e25b01287)

**Comment by carmenfan** [comment]
> @carmenfan
> 
> With the floating precision values not being converted to string and back, we get more decimal places in the frontend (see pictures below). Sebastian thinks the user might not expect that many decimals and that maybe adding a teamspace setting would be justified to cut them off at a user-defined length. What do you think?
> 
> After the changes: ![image](https://private-user-images.githubusercontent.com/18236690/358187856-e6b7e426-b7bb-4eba-9191-75bb50bd1bfa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM3MzgyOTcsIm5iZiI6MTcyMzczNzk5NywicGF0aCI6Ii8xODIzNjY5MC8zNTgxODc4NTYtZTZiN2U0MjYtYjdiYi00ZWJhLTkxOTEtNzViYjUwYmQxYmZhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODE1VDE2MDYzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI5NGZhNTBkNjRlZmY4YzE4NWM4YjMwMTAxYjMxN2QzYmRkNGVmOWRjNjEwMzFkMTQ1NGE0MTFhNjc3ODI0ZmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.wBnWbMnvp4DL7rt5h6YWB4GlE1lj35DqLIDSb1xudPc)
> 
> Before the changes: ![image](https://private-user-images.githubusercontent.com/18236690/358187961-a82e504e-aa7e-4566-8187-e30e25b01287.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM3MzgyOTcsIm5iZiI6MTcyMzczNzk5NywicGF0aCI6Ii8xODIzNjY5MC8zNTgxODc5NjEtYTgyZTUwNGUtYWE3ZS00NTY2LTgxODctZTMwZTI1YjAxMjg3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODE1VDE2MDYzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRkNWYxZDExMzYxNzRmNGYxOTljN2I0ZmQyYWYwYzNmNmNhYzUxMWRiOGZmNGZlZjdiYmEzYTc5NjkwOGM4MmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.T44VGXxkYe6vlE2Ix3mmImI33AIyJcq9Y6-fWr43SEU)

@FJThiel I think it depends what the user sees in the original file. e.g. if you open it in navis and look for the same field to you get this amount of precision?

**Comment by sebjf** [comment]
@carmenfan @FJThiel, (just in case it wasn't clear!) my thoughts are that now we are storing floats as floats, we can choose what precision to show them at in the BIM card. That is, the metadata API would return the full precision (and appropriate encoding for any other standard types, like Date), and then the frontend would display appropriately. 
Therefore, we need to pick an appropriate precision, and we may want it to be configurable...
(As otherwise we'd need to choose the precision when building the metadata API response, and it seems odd to me to curtail the precision of such a type in what is inherently a machine-to-machine message, though perhaps we may need to if doing otherwise will break users scripts & integrations...)

**Comment by carmenfan** [comment]
@sebjf just to make sure I 100% got what you mean, you're suggesting to leave the number as is until we display it right? so what you're suggesting here is a front end job?

If that's what you mean then agreed, and we should probably check what revit/navis etc do to the number on their UI and just match it

**Comment by sebjf** [comment]
@carmenfan, yes thats what I meant!

**Comment by FJThiel** [comment]
@carmenfan @sebjf 
There might be another front end job as well with dates:
The NWD format can hand us date/time objects and now we are able to pass them as such to the database. However, this means that they look a bit less accessible in the frontend.
![image](https://github.com/user-attachments/assets/c553457c-3b22-4019-ab5c-b830ed625753)

**Comment by carmenfan** [comment]
@FJThiel do we get date formats for any other formats you've checked through?

**Comment by FJThiel** [comment]
@carmenfan 

No. The formats I get from Assimp and Revit don't support it, DWG has hardly any metadata to begin with, and Synchro, BIM, and DGN arrive as strings already at the bouncer. IFC does have a field for time, but it is a double and can have varying units depending on the project settings, so I am inclined to think that it does not necessarily include date/time and might be more for representing duration.

**Comment by FJThiel** [comment]
Bools are also currently not displayed in the frontend.

**Comment by FJThiel** [comment]
@sebjf @carmenfan 

What is our line on empty fields? For the NWD files, I had replaced the old OdNwProperty with the newer OdNwDataProperty because it is neater and the old one will be removed by ODA soon. However, this changed the behaviour of the magic ODA state machine in the background. Previously, we would only get meta data fields such as Item::Name, Item::Type, and Item::InternalType if they actually have values set. The new behaviour appears to always give these fields and just fill them with "" if the node from the imported file has no value for that field.

My comparison script obviously does not like it, and it will lead to some empty fields in the viewer, but apart from that I don't see any issues with the processing of the file.

What is more important to us here? That we accurately represent in the DB and frontend what the file and importer give us (even if it means empty fields) or do we want to cull that before it makes it in the database?

**Comment by carmenfan** [comment]
I think if it's empty we should ignore it as if the field doesn't exist
